src/S3/S3Transfer/MultipartCopier.php

```diff
@@ -0,0 +1,230 @@
+use Aws\Arn\ArnParser;
+use Aws\Arn\AccessPointArn;
+use GuzzleHttp\Promise\Create;
+use GuzzleHttp\Promise\PromiseInterface;
+use Aws\CommandPool;
+use Throwable;
+
+/**
+ * Multipart copier implementation.
+ *
+ * Handles copying large objects between S3 locations using multipart copy.
+ *
+ * @property array $source Location of the data in the format of an array with:
+ *        - Bucket - Source bucket name
+ *        - Key - Source object key
+ *        - VersionId - Optional source version ID
+ */
+class MultipartCopier extends AbstractMultipartUploader
+{
+    /** @var array */
+    private array $source;
+
+    /**
+     * @param S3ClientInterface $s3Client
+     * @param array $createMultipartArgs
+     * @param array $config
+     * @param array $source
+     * @param string|null $uploadId
+     * @param array $parts
+     * @param TransferProgressSnapshot|null $currentSnapshot
+     * @param TransferListenerNotifier|null $listenerNotifier
+     */
+    public function __construct
+    (
+        S3ClientInterface $s3Client,
+        array $createMultipartArgs,
+        array $config,
+        array $source,
+        ?string $uploadId = null,
+        array $parts = [],
+        ?TransferProgressSnapshot $currentSnapshot = null,
+        ?TransferListenerNotifier $listenerNotifier = null,
+    ) {
+
+        if (empty($source['Bucket']) || empty($source['Key'])) {
+            throw new \InvalidArgumentException(
+                "The source array must contain 'Bucket' and 'Key' parameters"
+            );
+        }
+
+        if ($source['Bucket'] === $createMultipartArgs['Bucket']
+            && $source['Key'] === $createMultipartArgs['Key']) {
+            throw new \InvalidArgumentException(
+                "Source and destination cannot be the same object"
+            );
+        }
+        parent::__construct(
+            s3Client: $s3Client,
+            createMultipartArgs: $createMultipartArgs,
+            config: $config,
+            uploadId: $uploadId,
+            parts: $parts,
+            currentSnapshot: $currentSnapshot,
+            listenerNotifier: $listenerNotifier
+        );
+
+        $this->source = $source;
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    public function copy(): PromiseInterface
+    {
+        try {
+            $result = $this->promise()->wait();
+            return Create::promiseFor($result);
+        } catch (Throwable $e) {
+            return Create::rejectionFor($e);
+        }
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    protected function processMultipartOperation(): PromiseInterface
+    {
+        return $this->copyParts();
+    }
+
+    /**
+     * @return int
+     */
+    protected function getTotalSize(): int
+    {
+        if (! isset($this->calculatedObjectSize)) {
+            if (isset($this->config['object_size'])) {
+                $this->calculatedObjectSize = (int) $this->config['object_size'];
```

Comment:
nit: space between bang operator and isset call.  Also, I think you could combine the two if statements like 

```php
if (!isset($this->calculatedObjectSize)
    && isset($this->config['object_size'])
) {
    $this->calculatedObjectSize = (int) $this->config['object_size'];
} else ...